BUG: More robust FASTQ validation

thermokarst opened the pull request:
Fixes #148

Pair-programmed with @ebolyen

jairideout commented on q2_types/per_sample_sequences/_format.py:
```diff
@@ -105,25 +107,37 @@ class FastqGzFormat(model.BinaryFileFormat):
+                    raise ValidationError('Missing sequence for record '
+                                          'beginning on line %d'
+                                          % (i * 4 + 1))
+                elif not seq.isupper():
```
Since `.isupper()` takes a linear pass, I wonder if we could validate the sequence using `skbio.DNA(seq, validate=True)` instead? If the performance is comparable it'd probably be worth having the extra validation.

ebolyen replied:
DNA's constructor is pretty slow and Python str operations are super fast, our reasoning was that it's pretty unlikely that a user would have a fastq file with the wrong characters which didn't also cause the `@` check to throw up. And a protein fastq doesn't really make sense. Is it typical that the sequence's characters are the issue with fastq?

ebolyen replied:
Just tested:
```python
In [1]: import skbio

In [2]: a = "ACGT" * 50

In [3]: %timeit a.isupper()
The slowest run took 4.55 times longer than the fastest. This could mean that an intermediate result is being cached.
1000000 loops, best of 3: 1.1 µs per loop

In [4]: %timeit skbio.DNA(a, validate=True)
The slowest run took 11.97 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 3: 15.9 µs per loop
```

jairideout replied:
Performance hit doesn't seem worth it. Thanks for profiling!

jairideout commented on q2_types/per_sample_sequences/_format.py:
```diff
@@ -105,25 +107,37 @@ class FastqGzFormat(model.BinaryFileFormat):
+                    raise ValidationError('Lowercase case sequence on line %d'
+                                          % (i * 4 + 2))
+
+                if sep is None:
```
Do you think it's worth checking that the separator line starts with `+`?

ebolyen replied:
The `@` should trip up in the next record, so we thought we could save an operation in this loop.

jairideout replied:
I mean for validation purposes (it's part of the fastq spec)

ebolyen replied:
Following up from slack. It's possible for *something else* to be here, which would trip things up. The FASTQ header is allowed to be repeated, so we should use `sep.startswith('+')` as an `else` block.

jairideout commented on q2_types/per_sample_sequences/_format.py:
```diff
@@ -105,25 +107,37 @@ class FastqGzFormat(model.BinaryFileFormat):
+                                          'beginning on line %d'
+                                          % (i * 4 + 1))
+
+                if qual is None:
```
One quick validation that might be useful: check that the lengths of `seq` and `qual` are equal (i.e. each base has a corresponding quality score character).

ebolyen replied:
:+1: That could catch some situations where the fastq is cat-ed together.

ebolyen commented:
merge conflicts 😿

thermokarst commented:
Thanks @jairideout! I have addressed your recommendations in a [squash commit](https://github.com/qiime2/q2-types/pull/153/commits/3128f1ada501a71d2dfb37d15a6db47a47348595) 🎃. I also fixed the merge conflict.
